---
title: "**pradMAEO**: The **PR**ostate **AD**enocarcinoma
**M**ulti**A**ssay**E**xperiment **O**bject using data from _RTCGAToolbox_"
author:
- name: Lucas Schiffer
  affiliation: CUNY Graduate School of Public Health and Health Policy
- name: Marcel Ramos
  affiliation: CUNY Graduate School of Public Health and Health Policy
date: "`r BiocStyle::doc_date()`"
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: yes
abstract: >
  This vignette generates a
  MultiAssayExperiment object from TCGA (The Cancer Genome Atlas) data using 
  Specifically, the `getFirehoseData()` method of the
  `r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package (the fork available
  from https://www.github.com/LiNk-NY/RTCGAToolbox) is used to access and read
  in data; output is then further coerced to fit the MultiAssayExperiment object
  specifications. 
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{MultiAssayExperiment: Prostate Cancer Data}
  %\VignetteEncoding{UTF-8}
---

# Prerequisites

```{r, include = FALSE}
knitr::opts_knit$set(eval = FALSE)
if (!require(TCGAutils, quietly = TRUE)) {
    BiocInstaller::biocLite("waldronlab/TCGAutils", suppressAutoUpdate = TRUE,
        suppressUpdates = TRUE)
}
suppressPackageStartupMessages({
    library(RTCGAToolbox)
    library(TCGAutils)
})
```

Note: This vignette *requires* `TCGAutils` which currently only available on Github.

```{r, eval=FALSE}
library(RTCGAToolbox)
## Download and install GitHub resource
if (!require(TCGAutils, quietly = TRUE)) {
    BiocInstaller::biocLite("waldronlab/TCGAutils", suppressAutoUpdate = TRUE,
        suppressUpdates = TRUE)
}
library(TCGAutils)
```

These and other packages available in Bioconductor or CRAN are loaded as
follows.

```{r}
# BIOCONDUCTOR
library(MultiAssayExperiment)
library(RaggedExperiment)
```

# Argument Definitions

The `r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package provides the
`getFirehoseDatasets()` method for obtaining the names of all 33 cohorts
contained within the TCGA data. Beyond the 33 cohorts, there are 5 additional
"pan" cohorts where data of multiple cohorts was merged - information about the
cohorts is available via the TCGA
[website](http://cancergenome.nih.gov/cancersselected). Additionally, the
`getFirehoseRunningDates()` and `getFirehoseAnalyzeDates()` methods are used
to obtain the most recent running and analysis dates.

```{r}
dataset <- getFirehoseDatasets()[27] # PRAD
stopifnot(identical(dataset, "PRAD"))
runDate <- getFirehoseRunningDates(last = 1)
analyzeDate <- getFirehoseAnalyzeDates(last = 1)
```

# Function Definition

A function, `buildMultiAssayExperiment()`, is defined as shown below for the
purpose of creating a new MultiAssayExperiment object with a single line of
code. It accepts the arguments defined in the above chunk. It works for a single
dataset at a time.

In the first part of the function, the existence of the data directory is
checked and it is created if necessary. Then a cohort object is either loaded
or serialized from the `getFirehoseData()` method and saved to the data
directory. Once serialized, `colData` is extracted from the clinical slot and
the rownames are cleaned by `gsub()` and `type.convert()` functions. 

A named list of extraction targets is then created from the slot names of the
cohort object and the `biocExtract()` function is used within a `try` call.
The outputs are then passed to `generateMap()` which will generate a `sampleMap`
specific to TCGA data.

Finally, the named `ExperimentList` of extracted datasets, the `colData`, and
the generated sample map can be passed to the `MultiAssayExperiment()`
constructor function. The constructor function will ensure that orphaned
samples, samples that don't match a record in `colData`, are removed.
A `MultiAssayExperiment` will be created, serialized and saved to the data
directory, making it easier to return to in the future.

```{r}
buildMultiAssayExperiment <- function(TCGAcode, runDate, analyzeDate, dataType=
    c("RNASeqGene", "RNASeq2GeneNorm", "miRNASeqGene", "CNASNP", "CNVSNP",
    "CNASeq", "CNACGH", "Methylation", "mRNAArray", "miRNAArray", "RPPAArray",
    "Mutation", "GISTIC"), dataDirectory = "/tmp", force = FALSE) {
        if (!dir.exists(dataDirectory))
            dir.create(dataDirectory)
    
    message("\n######\n", "\nProcessing ", TCGAcode, " : )\n", "\n######\n")
    serialPath <- file.path(dataDirectory, paste0(TCGAcode, ".rds"))
    
    targets <- c("RNASeqGene", "RNASeq2GeneNorm", "miRNASeqGene",
                     "CNASNP", "CNVSNP", "CNASeq", "CNACGH", "Methylation",
                     "mRNAArray", "miRNAArray", "RPPAArray", "Mutation",
                     "GISTIC")
    names(targets) <- targets
    
    dataType <- match.arg(dataType, targets, several.ok = TRUE)
    datanames <- dataType
    dataType <- rep(TRUE, length(dataType))
    names(dataType) <- datanames
    
    if (file.exists(serialPath)) {
        cancer.object <- readRDS(serialPath)
    } else {
        cancer.object <- do.call(getFirehoseData,
            args = c(list(dataset = TCGAcode,
                runDate = runDate,
                gistic2Date = analyzeDate,
                RNAseqNorm = "raw_counts",
                RNAseq2Norm =
                    "normalized_count",
                forceDownload = FALSE,
                destdir = dataDirectory,
                fileSizeLimit = 500000,
                getUUIDs = FALSE), as.list(dataType))
            )
        # Uncomment to save to your directory
        # saveRDS(cancer.object, file = serialPath, compress = "bzip2")
    }
    ## Add clinical data from RTCGAToolbox
    clinical.data <- selectType(cancer.object, "clinical")
    rownames(clinical.data) <- toupper(gsub("\\.", "-",
                                            rownames(clinical.data)))
    clinical.data[] <- apply(clinical.data, 2, type.convert)
    
    dataList <- lapply(targets, function(datType) {
        tryCatch({biocExtract(cancer.object, datType)},
                 error = function(e) {
                     message(datType, " does not contain any data!")
                 })
    })
    data.full <- Filter(length, dataList)
    
    isList <- vapply(data.full, is.list, logical(1L))
    unlisted <- unlist(data.full[isList])
    data.full <- c(data.full[!isList], unlisted)
    newmap <- TCGAutils::generateMap(data.full, clinical.data,
        TCGAutils::TCGAbarcode)
    mae <- MultiAssayExperiment(data.full, clinical.data, newmap)
    return( mae )
}
```

# Function Call

Lastly, it is necessary to call the `buildMultiAssayExperiment()` function
defined above and pass it the arguments defined using the
`r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package. Using this function, a
`MultiAssayExperiment` object for the prostate
adenocarcinoma cohort (`PRAD`) is created with a single call.

```{r}
buildMultiAssayExperiment(TCGAcode=dataset, runDate=runDate,
    analyzeDate=analyzeDate, dataType = "miRNASeqGene", dataDirectory = "tmp")
```
