---
title: "**pradMAEO**: The **PR**ostate **AD**enocarcinoma
**M**ulti**A**ssay**E**xperiment **O**bject using data from _RTCGAToolbox_"
author: "Lucas Schiffer"
date: "`r BiocStyle::doc_date()`"
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: yes
abstract: >
  This vignette generates a
  MultiAssayExperiment object from TCGA (The Cancer Genome Atlas) data using 
  Specifically, the `getFirehoseData()` method of the
  `r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package (the fork available
  from https://www.github.com/LiNk-NY/RTCGAToolbox) is used to access and read
  in data; output is then further coerced to fit the MultiAssayExperiment object
  specifications. 
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{MultiAssayExperiment: Prostate Cancer Data}
  %\VignetteEncoding{UTF-8}
---

# Prerequisites

```{r, include = FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

Note that this vignette *requires* the LiNk-NY version of RTCGAToolbox, not the version on Bioconductor, and that TCGAutils is currently only available on Github.

```{r}
BiocInstaller::biocLite("LiNk-NY/RTCGAToolbox")
BiocInstaller::biocLite("waldronlab/TCGAutils")
```

These and other packages available in Bioconductor or CRAN are loaded as
follows.

```{r}
library(MultiAssayExperiment)
library(RTCGAToolbox)
library(TCGAutils)
library(readr)
library(RaggedExperiment)
```

# Argument Definitions

The `r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package provides the
`getFirehoseDatasets()` method for obtaining the names of all 33 cohorts
contained within the TCGA data. Beyond the 33 cohorts, there are 5 additional
"pan" cohorts where data of multiple cohorts was merged - information about the
cohorts is available via the TCGA
[website](http://cancergenome.nih.gov/cancersselected). Additionally, the
`getFirehoseRunningDates()` and `getFirehoseAnalyzeDates()` methods are used
to obtain the most recent running and analysis dates.

```{r}
dataset <- getFirehoseDatasets()[27] # PRAD
stopifnot(identical(dataset, "PRAD"))
runDate <- getFirehoseRunningDates()[1]
analyzeDate <- getFirehoseAnalyzeDates()[1]
```

# Function Definition

A function, `buildMultiAssayExperiment()`, is defined as shown below for the
purpose of creating a new MultiAssayExperiment object with a single line of
code. It accepts the arguments defined in the above chunk. It works for a single
dataset at a time.

In the first part of the function, the existence of the data directory is
checked and it is created if necessary. Then a cohort object is either loaded
or serialized from the `getFirehoseData()` method and saved to the data
directory. Once serialized, `colData` is extracted from the clinical slot and
the rownames are cleaned by `gsub()` and `type.convert()` functions. 

A named list of extraction targets is then created from the slot names of the
cohort object and the `TCGAextract()` function is used within a try statement.
The outputs are then passed to `generateMap()` which will generate a `sampleMap`
specific to TCGA data.

Finally, the named `ExperimentList` of extracted datasets, the `colData`, and
the generated sample map can be passed to the `MultiAssayExperiment()`
constructor function. The constructor function will ensure that orphaned
samples, samples that don't match a record in `colData`, are removed.
A `MultiAssayExperiment` will be created, serialized and saved to the data
directory, making it easier to return to in the future.

```{r}
buildMultiAssayExperiment <- function(TCGAcode, runDate, analyzeDate) {
    message("\n######\n", "\nProcessing ", TCGAcode, " : )\n", "\n######\n")
    serialPath <- file.path("data", paste0(TCGAcode, ".rds"))
    if (file.exists(serialPath)) {
        cancer.object <- readRDS(serialPath)
    } else {
        cancer.object <- getFirehoseData(TCGAcode, runDate = runDate,
                                         gistic2_Date = analyzeDate,
                                         RNAseq_Gene = TRUE,
                                         Clinic = TRUE,
                                         miRNASeq_Gene = TRUE,
                                         RNAseq2_Gene_Norm = TRUE,
                                         CNA_SNP = TRUE,
                                         CNV_SNP = TRUE,
                                         CNA_Seq = TRUE,
                                         CNA_CGH = TRUE,
                                         Methylation = FALSE,
                                         Mutation = TRUE,
                                         mRNA_Array = TRUE,
                                         miRNA_Array = TRUE,
                                         RPPA_Array = TRUE,
                                         RNAseqNorm = "raw_counts",
                                         RNAseq2Norm =
                                             "normalized_count",
                                         forceDownload = FALSE,
                                         destdir = "./tmp",
                                         fileSizeLimit = 500000,
                                         getUUIDs = FALSE)
        saveRDS(cancer.object, file = serialPath, compress = "bzip2")
    }
    ## Add clinical data from RTCGAToolbox
    clinical.data <- Clinical(cancer.object)
    rownames(clinical.data) <- toupper(gsub("\\.", "-",
                                            rownames(clinical.data)))
    clincal.data[] <- apply(clinical.data, 2, type.convert)
    
    ## slotNames in FirehoseData RTCGAToolbox class
    targets <- c("RNASeqGene", "RNASeq2GeneNorm", "miRNASeqGene",
                 "CNASNP", "CNVSNP", "CNAseq", "CNACGH", "Methylation",
                 "mRNAArray", "miRNAArray", "RPPAArray", "Mutations",
                 "gistica", "gistict")
    names(targets) <- targets
    dataList <- lapply(targets, function(datType) {
        tryCatch({TCGAutils::TCGAextract(cancer.object, datType)},
                 error = function(e) {
                     message(datType, " does not contain any data!")
                 })
    })
    data.full <- Filter(function(x) {!is.null(x)}, dataList)
    for (i in seq_along(data.full)){
        if(is(data.full[[i]], "GRangesList")){
            data.full[[i]] <- RaggedExperiment::RaggedExperiment(data.full[[i]])
        }
    }
    newmap <- generateMap(data.full, clinical.data, TCGAutils::TCGAbarcode)
    mae <- MultiAssayExperiment(data.full, clinical.data, newmap)
    return( mae )
}
```

# Function Call

Lastly, it is necessary to call the `buildMultiAssayExperiment()` function
defined above and pass it the arguments defined using the
`r BiocStyle::Githubpkg("LiNk-NY/RTCGAToolbox")` package. Using this function, a
`MultiAssayExperiment` object for the prostate
adenocarcinoma cohort (`PRAD`) is created with a single call.

```{r}
mae <- buildMultiAssayExperiment(TCGAcode=dataset, runDate=runDate,
                                  analyzeDate=analyzeDate)
```
