---
title: "pradMAEO - <strong>PR</strong>ostate <strong>AD</strong>enocarcinoma <strong>M</strong>ulti<strong>A</strong>ssay<strong>E</strong>xperiment <strong>O</strong>bject"
author: "Lucas Schiffer"
date: "January 8, 2016"
output: html_document
---

This vignette is the current working document for the development of a prostate adenocarcinoma MultiAssayExperiment object. Data is taken from The Cancer Genome Atlas (TCGA) using the getFirehoseData() method of the RTCGAToolbox package, then manipulated to fit into the MultiAssayExperiment object.

GitHub Source <- <a href="https://github.com/schifferl/biocMultiAssay/blob/master/vignettes/pradMAEO.Rmd">https://github.com/schifferl/biocMultiAssay/blob/master/vignettes/pradMAEO.Rmd</a><br>
Rpubs HTML <- <a href="https://rpubs.com/schifferl/pradMAEO">https://rpubs.com/schifferl/pradMAEO</a>

```{r, echo=FALSE, message=FALSE, warning=FALSE}
#install bioconductor
#source("https://bioconductor.org/biocLite.R")
#biocLite()

#install RTCGAToolbox
#BiocInstaller::biocLite("LiNk-NY/RTCGAToolbox")
library(RTCGAToolbox)

#install biocMultiAssay
#BiocInstaller::biocLite("vjcitn/biocMultiAssay")
library(biocMultiAssay)

#RTCGAToolbox help
#?RTCGAToolbox
```

Cohort datasets can be generated with the getFirehoseDatasets() method.

```{r}
getFirehoseDatasets()
```

Analysis and Running dates can be obtained through the getFirehoseAnalyzeDates() and getFirehoseRunningDates() methods of the RTCGAToolbox.

```{r}
getFirehoseAnalyzeDates()
getFirehoseRunningDates()
```

The <strong>PR</strong>ostate <strong>AD</strong>enocarcinoma dataset is identified by the cohort name "PRAD" and can be obtained using the getFirehouseData() method. Once returned the FirehoseData object can be serialized for manipulation. Use of the extract() method is necessary to convert data into a ExpressionSet or GRangesList objects, which can then be used to create a MuliAssayExperiment object.

```{r message=FALSE, warning=FALSE, results='hide'}
x <- "PRAD"
y <- getFirehoseRunningDates()[1]
z <- getFirehoseAnalyzeDates()[1]

if(file.exists("data/prad.Rda")) {
  load("data/prad.Rda")
} else {
  prad <- getFirehoseData(x, runDate = y, gistic2_Date = z, RNAseq_Gene = TRUE, Clinic = TRUE, 
          miRNASeq_Gene = TRUE, RNAseq2_Gene_Norm = TRUE, CNA_SNP = TRUE, CNV_SNP = TRUE, CNA_Seq = TRUE, 
          CNA_CGH = TRUE, Methylation = TRUE, Mutation = TRUE, mRNA_Array = TRUE, miRNA_Array = TRUE,
          RPPA = TRUE, RNAseqNorm = "raw_counts", RNAseq2Norm = "normalized_count", forceDownload = TRUE, 
          destdir = ".", fileSizeLimit = 500, getUUIDs = FALSE)
  save(prad, file = "data/prad.Rda") #comment when kniting html
}

names <- slotNames(prad)[5:17] #change for confilict

for(i in names) {
  try(assign(i, extract(prad, i, clinical = TRUE)), silent = TRUE)
}

vars <- ls()

clean <- c("vars", "clean")

for(i in vars) {
  if(!(i %in% names)) {
    clean <- c(clean, i)
  }
}

rm(list = clean)
```

The global environment should now contain the relevant data serialized as ExpressionSet or GRangesList objects.

```{r}
ls.str()
```

In order to create an EList object it is necessary to convert GRangesList objects to RangedRaggedAssay objects using the RangedRaggedAssay() method, such that assay methods for a MultiAssayExperiment are provided. Within the if block, the nameof variable is instantiated again after type conversion so that it references the RangedRaggedAssay rather than the GRangesList object.

```{r}
vars <- ls()
listof <- list()

for(i in vars) {
  nameof <- get(i)
  if(inherits(nameof, "GRangesList")) {
    assign(i, RangedRaggedAssay(nameof))
    nameof <- get(i)
  }
  listof <- c(listof, list(nameof))
}

experiments <- Elist(listof)

rm(vars, i, nameof, listof)
```
